vmm/src/memory_manager.rs

```diff
@@ -1477,9 +1462,121 @@ impl MemoryManager {
+                        let offset =
+                            page_size * ((i * pages_per_thread) + std::cmp::min(i, remainder));
+                        // SAFETY: FFI call with correct arguments
+                        let ret = unsafe {
+                            let addr = r.as_ptr().add(offset);
+                            libc::madvise(addr as _, pages * page_size, libc::MADV_POPULATE_WRITE)
+                        };
+                        if ret != 0 {
+                            let e = io::Error::last_os_error();
+                            warn!("Failed to prefault pages: {}", e);
+                        }
+                    });
+                }
+            });
+        }
+
+        if region.file_offset().is_none() && thp {
+            info!(
+                "Anonymous mapping at 0x{:x} (size = 0x{:x})",
+                region.as_ptr() as u64,
+                size
+            );
+            // SAFETY: FFI call with correct arguments
+            let ret = unsafe { libc::madvise(region.as_ptr() as _, size, libc::MADV_HUGEPAGE) };
+            if ret != 0 {
+                let e = io::Error::last_os_error();
+                warn!("Failed to mark pages as THP eligible: {}", e);
+            }
+        }
+
         Ok(Arc::new(region))
     }
 
+    // Duplicate of `memory_zone_get_align_size` that does not require a `zone`
+    fn get_prefault_align_size(
+        backing_file: &Option<PathBuf>,
+        hugepages: bool,
+        hugepage_size: Option<u64>,
+    ) -> Result<u64, Error> {
+        // SAFETY: FFI call. Trivially safe.
+        let page_size = unsafe { libc::sysconf(libc::_SC_PAGESIZE) as u64 };
+
+        // There is no backend file and the `hugepages` is disabled, just use system page size.
+        if !hugepages {
+            return Ok(page_size);
+        }
+
+        // The `hugepages` is enabled and the `hugepage_size` is specified, just use it directly.
+        if hugepages {
+            if let Some(hugepage_size) = hugepage_size {
+                return Ok(hugepage_size);
+            }
+        }
+
+        // There are two scenarios here:
+        //  - `hugepages` is enabled but `hugepage_size` is not specified:
+        //     Call `statfs` for `/dev/hugepages` for getting the default size of hugepage
+        //  - The backing file is specified:
+        //     Call `statfs` for the file and get its `f_bsize`.  If the value is larger than the page
+        //     size of normal page, just use the `f_bsize` because the file is in a hugetlbfs.  If the
+        //     value is less than or equal to the page size, just use the page size.
+        let path = backing_file
+            .as_ref()
+            .map_or(Ok("/dev/hugepages"), |pathbuf| {
+                pathbuf.to_str().ok_or(Error::InvalidMemoryParameters)
+            })?;
+
+        let align_size = std::cmp::max(page_size, statfs_get_bsize(path)?);
+
+        Ok(align_size)
```

Comment:
Agreed and updated. A note, I left`memory_zone_get_align_size` untouched, but it may make sense to refactor and combine `get_prefault_align_size` and `memory_zone_get_align_size` in a subsequent PR to reduce logic duplication.